engine/crates/runtime/src/fetch.rs

```diff
@@ -60,6 +72,58 @@ impl Fetcher {
             inner: Arc::new(fetcher),
         }
     }
+
+    #[allow(unused_variables)]
+    pub async fn post<SleepFut>(
+        &self,
+        request: &FetchRequest<'_>,
+        sleep: impl Fn(Duration) -> SleepFut + Send + Sync,
+    ) -> FetchResult<FetchResponse>
+    where
+        SleepFut: std::future::Future<Output = ()> + Send,
+    {
+        let rate_limit = request
+            .rate_limiter
+            .limit(&request.subgraph_name)
+            .map_err(FetchError::RateLimit);
+
+        let mut result = rate_limit.and_then(|_| self.inner.post(request)).await;
+
+        let Some(retry_budget) = request.retry_budget else {
+            return result;
+        };
+
+        let mut counter = 0;
+
+        loop {
+            match result {
+                Ok(bytes) => {
+                    retry_budget.deposit();
+                    return Ok(bytes);
+                }
+                Err(err) => {
+                    if retry_budget.withdraw().is_ok() {
+                        let jitter = rand::random::<f64>() * 2.0;
+                        let exp_backoff = (100 * 2u64.pow(counter)) as f64;
+                        let backoff_ms = (exp_backoff * jitter).round() as u64;
+
+                        sleep(Duration::from_millis(backoff_ms)).await;
+
+                        counter += 1;
+
+                        let rate_limit = request
```

Comment:
nitpick, but IMHO `rate_limit` variable isn't a great name. It's more an `am_i_rate_limited`. So at that point you could just do:
```rust
request..rate_limiter.limit(&request.subgraph_name).await?
self.inner.post(request).await
```
would be clearer imho

Reply:
but we need to get that one result from both futures.

Reply:
what do you mean? the rate limiter raises an error or not, you're using the ok value.

Reply:
ah it's loop, ok nevermind